enhancements/update/upgrade-without-registry.md

```diff
@@ -0,0 +1,893 @@
+check that there is at least twice the space indicated in the `size` field. If
+there is not enough space it will be reported as error conditions in the
+`CluterVersion` object and the upgrade process will be aborted.
+
+If there is enough space the `bundle-loader` pod will start a temporary embedded
+image registry server, listening in a randomly selected local port, using a self
+signed certificate to serve the images from the
+`/var/lib/upgrade/4.13.17-x86_64` directory. It will then node to trust the self
+signed certificate and configure CRI-O to use that temporary image registry
+server as a mirror for the release images, creating a
+`/etc/containers/registries.conf.d/upgrade-mirror.conf` file with a content
+similar to this:
+
+```toml
+[[registry]]
+prefix = "quay.io/openshift-release-dev/ocp-release"
+location = "quay.io/openshift-release-dev/ocp-release"
+
+[[registry.mirror]]
+location = "localhost:12345/openshift-release-dev/ocp-release"
+
+[[registry]]
+prefix = "quay.io/openshift-release-dev/ocp-v4.0-art-dev"
+location = "quay.io/openshift-release-dev/ocp-v4.0-art-dev"
+
+[[registry.mirror]]
+location = "localhost:12345/openshift-release-dev/ocp-v4.0-art-dev"
+```
+
+This configuration will be removed by the `bundle-loader` once the images have
+been successfully loaded into the `/var/lib/containers/storage` directory.
+
+Next the `bundle-loader` will start to load the images into the
+`/var/lib/containers/storage` directory. To do that it will use the gRPC of
+CRI-O to run the equivalent of `crictl pull` for each of the images. When that
+is completed the `bundle-laoder` will update the new
+`status.desired.bundle.nodes.*.loaded` field. For example, when the images have
+been loaded for `node0` the status will look like this:
+
+```yaml
+status:
+  desired:
+    bundle:
+      monitor: true
+      file: /dev/sdb
+      nodes:
+        node0:
+          extracted: true
+          loaded: true
+          metadata: '{ ... }'
+        node1:
+          extracted: true
+          loaded: false
+          metadata: '{ ... }'
+        node2:
+          extracted: true
+          loaded: false
+          metadata: '{ ... }'
+```
+
+At that point the `bundle-loader` job will finish.
+
+When all the nodes have the bundle loaded (the `loaded` fields of all nodes is
+`true`) the cluster version operator will check the new
+`spec.desiredUpdate.hold` field of the `ClusterVersion` object. This field will
+be used to indicate if the upgrade has to be started immediately, or if should
+wait till the user explicitly sets it to `false`. The default value for that
+will be `false`. This is intended for situations where the user wishes to
+prepare everything for the upgrade in advance, but wants to perform the upgrade
+later.
+
+When the value of the `spec.desiredUpdate.hold` field is `false` the cluster
+version operator will trigger the regular upgrade process, adding something
+like this to the `ClusterVersion` object:
```

Comment:
Having the CVO give itself instructions like this makes me think that it might be possible to create a new operator, outside of OCP, to manage this process. Did you consider that approach?

Reply:
Yes, we did consider that approach. Actually that is what we did in the [prototype](https://github.com/jhernand/upgrade-tool): a regular application independent of `oc adm upgrade ...` an independent of CVO. It works, as you can see in this [demo](https://drive.google.com/file/d/1QvFA1-VsMi1CsWMuIiV_nFdA3Giz3SGd) but I believe the user experience and the supportability would be much better if it is part of OCP.

Reply:
Why would it be easier to support in the CVO?

Reply:
It would be easier to support if it is part of OCP because it would be tested and released as part of OCP, customers will not have to open support tickets for a different product and team, and support engineers would follow the same processes that they follow for any other OCP ticket, escalating to the OCP engineering team if needed.

That doesn't mean that it needs to be part of CVO, although that looks like the natural place to me, it could be a separate operator, but still part of OCP.